whylog/config/parsers.py

```diff
@@ -57,3 +60,93 @@ def create_from_intent(cls, parser_intent):
     @classmethod
     def from_dao(cls, serialized_parser):
         return RegexParser(**serialized_parser)
+
+
+class ConcatedRegexParser(object):
+    NO_MATCH = frozendict()
+
+    def __init__(self, parser_list):
+        self._parsers = parser_list
+        forward, backward = self._create_concated_regexes()
+        self._forward_regex = re.compile(forward)
+        self._backward_regex = re.compile(backward)
+        self._forward_parsers_indexes = self._get_indexes_of_groups_for_parsers(self._parsers)
+        self._backward_parsers_indexes = self._get_indexes_of_groups_for_parsers(
+            reversed(
+                self._parsers
+            )
+        )
+        self._numbers_in_list = self._number_in_list()
+
+    def _create_concated_regexes(self):
+        return (
+            "|".join(
+                "(" + parser.regex_str + ")" for parser in self._parsers
+            ), "|".join(
+                "(" + parser.regex_str + ")" for parser in reversed(self._parsers)
+            )
+        )
+
+    def _get_indexes_of_groups_for_parsers(self, parser_list):
+        indexes_dict = {}
+        free_index = 0
+        for parser in parser_list:
+            amount_of_group = parser.regex_str.count('(') - parser.regex_str.count('\(')
+            indexes_dict[parser.name] = (free_index, amount_of_group)
+            free_index += amount_of_group + 1
+        return indexes_dict
+
+    def _number_in_list(self):
+        regex_numbers = {}
+        number = 0
+        for parser in self._parsers:
+            regex_numbers[parser.name] = number
+            number += 1
+        return regex_numbers
+
+    def get_extracted_regex_params(self, line):
+        forward_matched = self._forward_regex.match(line)
+        backward_matched = self._backward_regex.match(line)
+        if forward_matched is None:
+            return ConcatedRegexParser.NO_MATCH
+        clues = {}
+        forward_groups = forward_matched.groups()
+        backward_groups = backward_matched.groups()
+        for name, indexes in self._forward_parsers_indexes.items():
```

Comment:
Iterate over `forward_groups` and check hits with the parsers index. This will reduce random memory access and improve readability.